### PR TITLE
istep 18 implementation

### PR DESCRIPTION
**Pull request by IgorBagnucki**

### CODE REVIEW

**Comment by macpijan** [comment]
@krystian-hebel Please review

File: src/include/cpu/power/istep_13.h [hunk 1]
```diff
@@ -311,3 +311,5 @@ void istep_13_11(void);
 void istep_13_13(void);
 void istep_14_1(void);
 void istep_14_5(void);
+void istep_18_11(void);
+void istep_18_12(void);
```
Comment (krystian-hebel) [inline src/include/cpu/power/istep_13.h]: Why are those here?

File: src/soc/ibm/power9/Makefile.inc [hunk 2]
```diff
@@ -20,6 +20,8 @@ romstage-y += istep_14_1.c
 romstage-y += istep_14_2.c
 romstage-y += istep_14_3.c
 romstage-y += istep_14_5.c
+romstage-y += istep_18_11.c
+romstage-y += istep_18_12.c
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/Makefile.inc]: This code should run in ramstage, not romstage.

File: src/soc/ibm/power9/istep_18_11.c [hunk 3]
```diff
@@ -0,0 +1,75 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#include <console/console.h>
+#include <cpu/power/istep_18.h>
+
+// TODO: this one will cange if we add more than one processor
+#define MDMT (1)
+
+static uint32_t calculate_topology_dealy(void)
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/istep_18_11.c]: Should this be `calculate_topology_delay`?

File: src/soc/ibm/power9/romstage.c [hunk 4]
```diff
@@ -363,6 +363,8 @@ void main(void)
 	// istep_14_3();
 	report_istep(14,4);	// no-op
 	istep_14_5();
+	istep_18_11();
+	istep_18_12();
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/romstage.c]: Move these to ramstage (e.g. `enable_soc_dev` in chip.c)

File: src/include/cpu/power/istep_18.h [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+// [54] RWX RX_TTYPE_4_DATA_PARITY_ERROR_MASK: Error mask: data parity error on received TType-4 from register 0x21.
+// [55] RWX PIB_MASTER_REQUEST_ERROR_MASK: Error mask: PIB master; request while busy error.
+// [56] RWX PIB_RESET_DURING_PIB_ACCESS_ERROR_MASK: Error mask: PIB reset received during a PIB master or PIB slave operation.
+// [57] RWX EXTERNAL_XSTOP_ERROR_MASK: Error mask: TOD received an external checkstop.
+// [58] RWX SPARE_ERROR_MASK_58: Error mask: spare.
+// [59] RWX SPARE_ERROR_MASK_59: Error mask: spare.
+// [60] RWX SPARE_ERROR_MASK_60: Error mask: spare.
+// [61] RWX SPARE_ERROR_MASK_61: Error mask: spare.
+// [62] RWX OSCSWITCH_INTERRUPT_MASK: Error mask: Interrupt from the oscillator switch.
+// [63] RWX SPARE_ERROR_MASK_63: Error mask: spare.
```
Comment (krystian-hebel) [inline src/include/cpu/power/istep_18.h]: Those are masks for the same bits as in the previous register, there is no need to describe every bit. One sentence that these correspond to the above should be enough.

File: src/include/cpu/power/istep_18.h [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+#define PERV_TOD_ERROR_MASK_REG_RX_TTYPE_0 (38)
+#define PERV_TOD_ERROR_MASK_REG_RX_TTYPE_1 (39)
+#define PERV_TOD_ERROR_MASK_REG_RX_TTYPE_2 (40)
+#define PERV_TOD_ERROR_MASK_REG_RX_TTYPE_3 (41)
+#define PERV_TOD_ERROR_MASK_REG_RX_TTYPE_4 (42)
+#define PERV_TOD_ERROR_MASK_REG_RX_TTYPE_5 (43)
+
+#define PERV_TOD_ERROR_REG_RX_TTYPE_2 (40)
+#define PERV_TOD_ERROR_REG_RX_TTYPE_4 (42)
+#define PERV_TOD_ERROR_REG_RX_TTYPE_5 (43)
```
Comment (krystian-hebel) [inline src/include/cpu/power/istep_18.h]: These have the same values as above and they are used in close proximity of each other, it would make sense to keep just one of them (shorter).

File: src/include/cpu/power/istep_18.h [hunk 7]
```diff
@@ -0,0 +1,634 @@
...
+
+#define PERV_TOD_FSM_REG_IS_RUNNING (4)
+#define PERV_TOD_LOAD_TOD_MOD_REG_FSM_TRIGGER (0)
+#define PERV_TOD_START_TOD_REG_FSM_TRIGGER (0)
+#define PERV_TOD_TX_TTYPE_2_REG_TRIGGER (0)
+#define PERV_TOD_TX_TTYPE_4_REG_TRIGGER (0)
+#define PERV_TOD_TX_TTYPE_5_REG_TRIGGER (0)
```
Comment (krystian-hebel) [inline src/include/cpu/power/istep_18.h]: If the layout of those registers is the same, do not define a separate macro for each, one is enough.

File: src/soc/ibm/power9/istep_18_11.c [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+static int64_t calculate_m_path(uint64_t value)
+{
+  value |=
+    (read_scom(PERV_ROOT_CTRL8_SCOM) & P9A_PERV_ROOT_CTRL8_TP_PLL_CLKIN_SEL9_DC)
+    ? PPC_BIT(PERV_TOD_M_PATH_CTRL_REG_STEP_CREATE_DUAL_EDGE_DISABLE)
+    : 0;
+  if(MDMT)
+  {
+    value &=
+      ~PPC_BIT(0) & ~PPC_BIT(2) & ~PPC_BIT(13)
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/istep_18_11.c]: No names?

File: src/include/cpu/power/istep_18.h [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+// [24]    RW S_PATH_REMOTE_SYNC_ERROR_DISABLE: Slave path-01: Remote sync: Error disable.
+//         0 = Enable remote sync error as a step error.
+//         1 = Do not enable remote sync error as a step error.
+// [25]    RW S_PATH_REMOTE_SYNC_CHECK_M_CPS_DISABLE: Slave path-01: Remote sync: sync-step check: disable use of master path CPS.
+//         0 = Use master path CPS.
+//         1 = Do not use master path CPS.
+// [26:27] RW S_PATH_REMOTE_SYNC_CHECK_CPS_DEVIATION_FACTOR: Slave path-01: remote sync: sync-step check: CPS deviation factor.
+// [28:31] RW S_PATH_REMOTE_SYNC_CHECK_CPS_DEVIATION: Slave path-01: remote sync: sync-step check: CPS deviation.
+// [32:39] RW S_PATH_REMOTE_SYNC_MISS_COUNT_MAX: Slave path-01: remote sync: maximum of SYNC miss counts: 0 - 255 syncs.
+// [40:63] RO constant = 0
```
Comment (krystian-hebel) [inline src/include/cpu/power/istep_18.h]: This comment applies to whole file.

This is not the documentation, it should be enough to mention only the bits that are actually used. Sometimes just a name of the bit in the form of macro is self-explanatory. The documentation is publicly available, there is no need to copy it. Right now this file has ~600 lines that no one will ever need to read.

File: src/soc/ibm/power9/istep_18_11.c [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  return value;
+}
+
+void istep_18_11(void)
+{
+    printk(BIOS_EMERG, "starting istep 18.11\n");
+    report_istep(18,11);
+    write_scom(PERV_TOD_S_PATH_CTRL_REG, PPC_BITMASK(27, 32));
+    write_scom(PERV_TOD_PRI_PORT_0_CTRL_REG, calculate_topology_dealy());
+    write_scom(PERV_TOD_PSS_MSS_CTRL_REG, (MDMT ? PPC_BIT(1) : 0) | PPC_BIT(2) | PPC_BITMASK(9, 10));
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/istep_18_11.c]: No names?

File: src/soc/ibm/power9/istep_18_11.c [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    value |= PPC_BIT(1) | PPC_BIT(8) | PPC_BITMASK(14, 15);
+  }
+  return value;
+}
+
+void istep_18_11(void)
+{
+    printk(BIOS_EMERG, "starting istep 18.11\n");
+    report_istep(18,11);
+    write_scom(PERV_TOD_S_PATH_CTRL_REG, PPC_BITMASK(27, 32));
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/istep_18_11.c:65]: These are 3 separate fields in the register, please write it as such, preferably with a comment mentioning the names of those field (if we can't just use a macro).

Comment (krystian-hebel) [reply src/soc/ibm/power9/istep_18_11.c:65]: This applies to other non-zero registers as well.

File: src/soc/ibm/power9/istep_18_12.c [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+    write_scom(PERV_TOD_TX_TTYPE_2_REG, PPC_BIT(PERV_TOD_TX_TTYPE_2_REG_TRIGGER));
+    write_scom(PERV_TOD_LOAD_TOD_MOD_REG, PPC_BIT(PERV_TOD_LOAD_TOD_MOD_REG_FSM_TRIGGER));
+    write_scom(PERV_TOD_TX_TTYPE_5_REG, PPC_BIT(PERV_TOD_TX_TTYPE_5_REG_TRIGGER));
+    write_scom(PERV_TOD_LOAD_TOD_REG, PPC_BITMASK(50, 61));
+    write_scom(PERV_TOD_START_TOD_REG, PPC_BIT(PERV_TOD_START_TOD_REG_FSM_TRIGGER));
+    write_scom(PERV_TOD_TX_TTYPE_4_REG, PPC_BIT(PERV_TOD_TX_TTYPE_4_REG_TRIGGER));
+
+    // wait until TOD is running
+    wait_us(1000, read_scom(PERV_TOD_FSM_REG) & PPC_BIT(PERV_TOD_FSM_REG_IS_RUNNING));
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/istep_18_12.c]: Please add check for timeout.

File: src/soc/ibm/power9/istep_18_12.c [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+      | PPC_BIT(PERV_TOD_ERROR_MASK_REG_RX_TTYPE_4)
+      | PPC_BIT(PERV_TOD_ERROR_MASK_REG_RX_TTYPE_5));
+}
+
+void istep_18_12(void)
+{
+    printk(BIOS_EMERG, "starting istep 18.12\n");
+	report_istep(18,12);
+    init_tod_node();
+    printk(BIOS_EMERG, "ending istep 18.12\n");
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/istep_18_12.c]: Inconsistent indentation.

File: src/soc/ibm/power9/istep_18_11.c [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    uint32_t longest_topology_delay = 0;
+
+    return
+      - longest_topology_delay
+      + node_link_delay
+      + MDMT_TOD_GRID_CYCLE_STAGING_DELAY;
+  }
+
+  write_scom(PU_PB_ELINK_RT_DELAY_CTL_REG, PPC_BITMASK(2, 3));
+  uint64_t l_bus_mode_reg = read_scom(PU_PB_ELINK_DLY_0123_REG);
```
Comment (krystian-hebel) [inline src/soc/ibm/power9/istep_18_11.c:27]: Please use `_for_chiplet` versions here.

**Review by krystian-hebel — CHANGES_REQUESTED** [review]

File: src/include/cpu/power/istep_13.h (hunk 1 above)
Comment (IgorBagnucki) [reply src/include/cpu/power/istep_13.h]: Removed

File: src/soc/ibm/power9/Makefile.inc (hunk 2 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/Makefile.inc]: Changed to ramstage

File: src/soc/ibm/power9/istep_18_11.c (hunk 3 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/istep_18_11.c]: Fixed

File: src/include/cpu/power/istep_18.h (hunk 5 above)
Comment (IgorBagnucki) [reply src/include/cpu/power/istep_18.h]: Removed the comment and left only a note, that it is the same as previous one

File: src/include/cpu/power/istep_18.h (hunk 6 above)
Comment (IgorBagnucki) [reply src/include/cpu/power/istep_18.h]: Removed most of the entries, leaving only one

File: src/include/cpu/power/istep_18.h (hunk 7 above)
Comment (IgorBagnucki) [reply src/include/cpu/power/istep_18.h]: Fixed

File: src/include/cpu/power/istep_18.h (hunk 9 above)
Comment (IgorBagnucki) [reply src/include/cpu/power/istep_18.h]: Removed most of the comments, leaving only information about used bit fields.

File: src/soc/ibm/power9/istep_18_12.c (hunk 12 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/istep_18_12.c]: Added a check for timeout with an error message

File: src/soc/ibm/power9/istep_18_12.c (hunk 13 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/istep_18_12.c]: Fixed

File: src/soc/ibm/power9/istep_18_11.c (hunk 14 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/istep_18_11.c:27]: @krystian-hebel This line of code is the equivalent of a [statement bellow](https://github.com/open-power/hostboot/blob/46a489ba7864b1d32c4b37a1223a39f33ad12f66/src/import/chips/p9/procedures/hwp/nest/p9_tod_setup.C#L1332).
```
FAPI_TRY(fapi2::getScom(*(i_tod_node->i_target),
                            l_bus_mode_addr,
                            l_bus_mode_reg),
             "Error from getScom (0x%08X)!", l_bus_mode_addr);
```
`i_target` has a type of `fapi2::TARGET_TYPE_PROC_CHIP`. [source](https://github.com/open-power/hostboot/blob/46a489ba7864b1d32c4b37a1223a39f33ad12f66/src/import/chips/p9/procedures/hwp/nest/p9_tod_utils.H#L313)
```
struct tod_topology_node
{
    fapi2::Target<fapi2::TARGET_TYPE_PROC_CHIP>* i_target;
    bool i_tod_master;
    bool i_drawer_master;
    p9_tod_setup_bus i_bus_rx; // Current node's bus from which step/sync is
    //   received ("Receive Port" in eiinfo)
    p9_tod_setup_bus i_bus_tx; // Upstream node's bus from which step/sync
    //   is transmitted ("Drive Port" in eiinfo)
    std::list<tod_topology_node*> i_children;
    p9_tod_setup_conf_regs o_todRegs;
    uint32_t o_int_path_delay;
};
```
Therefore it is not chiplet id, but **processor id**. For the time being we support only one processor, but I agree, this code could be future proofed. However looking in the documenatation or our own code I can't find any information about access to the other-processor's scom adress except for [this note](https://github.com/3mdeb/openpower-coreboot-docs/blob/main/devnotes/scom_access.md#assembly-instruction-ldcix) which i'm not sure if is correct.

If You know how to access the correct processor's scom, please let me know and I'll fix the discussed code.

File: src/soc/ibm/power9/romstage.c (hunk 4 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/romstage.c]: Moved

File: src/soc/ibm/power9/istep_18_11.c (hunk 8 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/istep_18_11.c]: Fixed

File: src/soc/ibm/power9/istep_18_11.c (hunk 10 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/istep_18_11.c]: Fixed

File: src/soc/ibm/power9/istep_18_11.c (hunk 11 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/istep_18_11.c:65]: Fixed

File: src/soc/ibm/power9/istep_18_11.c (hunk 14 above)
Comment (IgorBagnucki) [reply src/soc/ibm/power9/istep_18_11.c:27]: @krystian-hebel is this solution correct for You, or should we prepare the code somehow for a second processor in topology?

Comment (krystian-hebel) [reply src/soc/ibm/power9/istep_18_11.c:27]: Sorry for late response, I was trying to understand what happens in Hostboot for PROC target, but this is messed up... Leave it as it is right now, it will need a new function for second processor anyway. I think this is the first step we  are implementing that doesn't start with "for each functional PROC" loop at the top level.

File: src/soc/ibm/power9/istep_18_11.c (hunk 11 above)
Comment (krystian-hebel) [reply src/soc/ibm/power9/istep_18_11.c:65]: Not fixed, there is no [27:32] field... This is also different than in analysis: https://github.com/3mdeb/openpower-coreboot-docs/commit/152c812916143cc49d4a64a491507745dea5ba55#diff-77adf17c2fd6a3eaecd5778d857416b7d295a5fa442e2462856c8f83bb16ca69R257
